public/docs/dart/latest/guide/forms.jade

```diff
@@ -0,0 +1,662 @@
+  1. Click inside the name box, then click outside it.
+  1. Add slashes to the end of the name.
+  1. Erase the name.
+
+  We'll see the following four sets of class names and their transitions:
+figure.image-display
+  img(src="/resources/images/devguide/forms/ng-control-class-changes.png" alt="Invalid Form")
+
+:marked
+  The (`ng-valid` | `ng-invalid`) pair are the most interesting to us, because we want to send a
+  strong visual signal when the values are bad. We also want to mark required fields.
+
+  We can do both at the same time with a colored bar on the left of the input box:
+
+figure.image-display
+  img(src="/resources/images/devguide/forms/validity-required-indicator.png" alt="Invalid Form")
+
+:marked
+  We achieve this effect by adding two styles to a new `styles.css` file
+  (under `web/`).
+
++makeExample('forms/dart/angular2_forms-7.2-CSS-NGVALID/web/styles.css', null, 'web/styles.css')
+:marked
+  These styles select for the two Angular validity classes and the HTML 5 "required" attribute.
+
+  To add the styles to the app,
+  update the `<head>` of `index.html` to link to `styles.css`.
++makeExample('forms/dart/angular2_forms-7.2-CSS-NGVALID/web/index.html', 'styles', 'web/index.html')(format="linenums:6")
+
+
+:marked
+  ## Show and hide validation error messages
+
+  We can do better.
+
+  The Name input box is required. Clearing it turns the bar red. That says *something* is wrong but we
+  don't know *what* is wrong or what to do about it.
+  We can leverage the `ng-invalid` class to reveal a helpful message.
+
+  Here's the way it should look when the user deletes the name:
+figure.image-display
+  img(src="/resources/images/devguide/forms/name-required-error.png" alt="Name required")
+
+:marked
+  To achieve this effect we need to add two chunks of code:
+  1. A local template variable in the `<input>` tag.
+     <!-- TODO: link to (./template-syntax.html#local-vars)-->
+  1. The "*is required*" message in a nearby `<div>`.
+     We'll display this message only if the control is invalid.
+
+  Here's an example of adding an error message to the "name" input box:
+  <!-- - var stylePattern = { (\#name="form")|(.*div.*$)|(Name is required)/gm }; -->
+- var stylePattern = { otl: /(#name=&quot;form&quot;)|(.*div.*$)|(Name is required)/gm };
++makeExample('forms/dart/angular2_forms-7.3-HIDDEN-ALERT/lib/hero_form_component.html', 'name-with-error-msg', 'lib/hero_form_component.html', stylePattern)(format="linenums:5")
+:marked
+  We initialized the template local variable with the string "form"
+  (`#name="form"`).
+
+  Angular recognizes that syntax and sets the `name` variable
+  to the `Control` object identified by the `ng-control` directive that,
+  not coincidentally, we called "name".
+
+  We bind the `Control` object's `valid` property to the element's `hidden` property.
+  While the control is valid, the message is hidden;
+  if it becomes invalid, the message is revealed.
+<a id="ng-form"></a>
+.l-sub-section
```

Comment:
I suggest that you replaces this sub-section with the following
```
.l-sub-section
  :marked
    ### The NgForm directive
    Recall from the previous section that `ng-control` registered this input box with the
    `NgForm` directive as "name".

    We didn't add the **[`NgForm`](../api/core/NgForm-class.html) directive** explicitly.
    Angular added it surreptiously, wrapping it around the `<form>` element

    The `NgForm` directive  supplements the `form` element with additional features.
    It collects `Controls` (elements identified by an `ng-control` directive)
    and monitors their properties including their validity.
    It has its own `valid` property which is true only if every contained
    control is valid.

    In this example, we are pulling the "name" control out of its `controls` collection
    and assigning it to the template local variable so that we can
    access the control's properties  &mdash; such as the control's own `valid` property.
```